### PR TITLE
Trigger event `experimental__woocommerce_blocks-cart-add-item` in the SSR Product Button vis Interactivity API

### LINKED ISSUE #42924: Product Collection: trigger `experimental__woocommerce_blocks-cart-add-item` JS action from Add to Cart button
Usually, developers can listen for the `experimental__woocommerce_blocks-cart-add-item` event like so:

```js
wp.hooks.addAction(
  'experimental__woocommerce_blocks-cart-add-item',
  'my-namespace',
  ( args ) => { /* some functionality */ }
)
```

However, on the Product Collection block this isn't firing - [the actions for the button](https://github.com/woocommerce/woocommerce/blob/9e9f0341e28abb9d405b6c761502eb199407f24b/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx#L135) don't seem to include anything about dispatching the event. 

To reproduce:

1. Open the site and go to a page showing the Product Collection block
2. Open your JS console and add:
```js
wp.hooks.addAction(
  'experimental__woocommerce_blocks-cart-add-item',
  'my-namespace',
  ( args ) => { console.log( 'item added to cart' ); }
)
```
3. Add an item to your cart and notice that no message was shown in the console.

### PR DESCRIPTION
**Pull request by kmanijak**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

In [usages of experimental prefix docs](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md#usages-of-experimental-prefix) we describe the `experimental__woocommerce_blocks-cart-add-item` which is not fired in Products or Product Collection blocks.

The implementation of these events is in JS so they're fired from client-side rendered blocks (e.g. All Products block). However, with many blocks being server-side rendered that functionality is not covered. 

There's another set of actions like `woocommerce_add_to_cart` ([documentation](https://wp-kama.com/plugin/woocommerce/hook/woocommerce_add_to_cart)) which should be used, however, for backward compatibility, in this PR I'm adding `experimental__woocommerce_blocks-cart-add-item` to Products and Product Collection blocks until we take action to clean up and unify events.

There's no specific contract in terms of payload of the event. The `product` is passed as an event's payload in [JS code](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/block.tsx#L76). I'm passing `product` in its shape from Product Button's store.

Closes https://github.com/woocommerce/woocommerce/issues/42924

### How to test the changes in this Pull Request:

**Note:** this PR can no longer be tested in releases, as it was reverted in https://github.com/woocommerce/woocommerce/pull/43177.

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Go to Editor
2. Add Products (Beta) and Product Collection block
3. Save and go to frontend
4. Open developer tools in your browser
5. In console add:

```
wp.hooks.addAction(
  'experimental__woocommerce_blocks-cart-add-item',
  'my-namespace',
  ( product ) => { console.log( 'Item added to cart: ' + product.name ); }
)
```

6. Add a product to a cart using Products (Beta) and Product Collection blocks
7. Expected: In both cases you should get `Item added to cart: <<PRODUCT NAME>>` in the console

### Changelog entry

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

Product Button: Trigger event `experimental__woocommerce_blocks-cart-add-item` in Products and Product Collection blocks.

#### Comment 

</details>

### CODE REVIEW

File: plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx [hunk 1]
```diff
@@ -144,6 +146,14 @@ const { state } = store< Store >( 'woocommerce/product-button', {
 					quantityToAdd
 				);
 
+				// Add a manual trigger of the event for backward compatibility
+				// and symmetry (remove event is triggered properly from Cart and Mini Cart).
+				// We'll try to unify the events in future.
+				yield doAction(
+					`experimental__woocommerce_blocks-cart-add-item`,
+					product
```
Comment (gigitux) [inline plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx:155]: Is there a particular reason why you didn't use `dispatchStoreEvent`? ([example](https://github.com/woocommerce/woocommerce/blob/ddd877441c74da8f128811e08405f8ca8c681c42/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/block.tsx/#L75-L77)).

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for the fix! Is there a particular reason why you didn't use `dispatchStoreEvent`?

If not, I guess that we should use `dispatchStoreEvent` function.

File: plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx (hunk 1 above)
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx:154]: @gigitux

Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx:155]: @gigitux 
Yes, so the problem is that in order to use `dispatchStoreEvent` you need to use`useStoreEvents` which is a [hook based on `useCallback` hook](https://github.dev/woocommerce/woocommerce/blob/ddd877441c74da8f128811e08405f8ca8c681c42/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/block.tsx/). It cannot be used outside of React component context and in this case we don't satisfy that need.

I was thinking of creating a non-hook helper function that would be simply the same as `dispatchStoreEvent` wrapping `doAction` but considering we're thinking of unifying and potentially cleaning the events up I reckoned I can add a single call here.

What do you think? Should I prepare a helped function for this? Also, I'm going to create a followup issue to cleanup and unify the events.

Comment (gigitux) [reply plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx:155]: @kmanijak Good point! I suggest opening a new issue and adding a comment in the code that contains the link to the issue.

Thanks for taking care of this! LGTM!

**Review by gigitux — APPROVED** [review]
Thanks for taking care of this PR! LGTM!

File: plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx (hunk 1 above)
Comment (kmanijak) [reply plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/button/frontend.tsx:155]: Issue created: https://github.com/woocommerce/woocommerce/issues/42985 and referenced in the comment 👍

**Pull request merged**